### PR TITLE
[MLIR] Add a BlobAttr interface for attribute to wrap arbitrary content and use it as linkLibs for ModuleToObject

### PR DESCRIPTION
**Pull request by joker-eph**
This change allows to expose through an interface attributes wrapping content as external
resources, and the usage inside the ModuleToObject show how we will be able to provide
runtime libraries without relying on the filesystem.

### CODE REVIEW

File: mlir/lib/IR/BuiltinAttributes.cpp [hunk 1]
```diff
@@ -862,6 +862,7 @@ DenseArrayAttrImpl<T> DenseArrayAttrImpl<T>::get(MLIRContext *context,
   Type elementType = DenseArrayAttrUtil<T>::getElementType(context);
   auto rawArray = ArrayRef<char>(reinterpret_cast<const char *>(content.data()),
                                  content.size() * sizeof(T));
+  llvm::errs() << " sizeof(T) " << sizeof(T) << "\n";
```
Comment (fabianmcg) [inline mlir/lib/IR/BuiltinAttributes.cpp:865]:
```suggestion
```

File: mlir/unittests/Target/LLVM/SerializeNVVMTarget.cpp [hunk 2]
```diff
@@ -215,3 +218,72 @@ TEST_F(MLIRTargetLLVMNVVM,
...
+      linkedCallback);
+  for (auto gpuModule : (*module).getBody()->getOps<gpu::GPUModuleOp>()) {
+    std::optional<SmallVector<char, 0>> object =
+        serializer.serializeToObject(gpuModule, options);
+
+    // Verify that we correctly linked in the library: the external call is
+    // replaced by the definition.
+    ASSERT_TRUE(!linkedLLVMIR.empty());
+    ASSERT_TRUE(
+        StringRef(linkedLLVMIR).contains("define internal void @bar() {"));
```
Comment (fabianmcg) [inline mlir/unittests/Target/LLVM/SerializeNVVMTarget.cpp:285]: This is fragile to IR changes, let's use the API.

**Review by fabianmcg — COMMENTED** [review]
In general LGTM, see the comments. Also, cool addition!

File: mlir/lib/Target/LLVM/ModuleToObject.cpp [hunk 3]
```diff
@@ -89,22 +92,51 @@ ModuleToObject::loadBitcodeFile(llvm::LLVMContext &context, StringRef path) {
...
+      llvm::SMDiagnostic error;
+      ArrayRef<char> data = blobAttr.getData();
+      std::unique_ptr<llvm::MemoryBuffer> buffer =
+          llvm::MemoryBuffer::getMemBuffer(StringRef(data.data(), data.size()),
+                                           "blobLinkedLib",
+                                           /*RequiresNullTerminator=*/false);
+      std::unique_ptr<llvm::Module> mod =
+          getLazyIRModule(std::move(buffer), error, context);
+      if (mod) {
+        llvmModules.push_back(std::move(mod));
```
Comment (fabianmcg) [inline mlir/lib/Target/LLVM/ModuleToObject.cpp:127]: Also, there should be a call to `handleBitcodeFile`.

File: mlir/unittests/Target/LLVM/SerializeNVVMTarget.cpp (hunk 2 above)
Comment (joker-eph) [reply mlir/unittests/Target/LLVM/SerializeNVVMTarget.cpp:285]: True, but isn't this on the same level as all the FileCheck CHECK in LLVM?

Comment (fabianmcg) [reply mlir/unittests/Target/LLVM/SerializeNVVMTarget.cpp:285]: Yes, but it's slightly harder to debug, at least FileCheck shows you where it failed to match.

File: mlir/lib/Target/LLVM/ModuleToObject.cpp (hunk 3 above)
Comment (joker-eph) [reply mlir/lib/Target/LLVM/ModuleToObject.cpp:127]: Done!

File: mlir/unittests/Target/LLVM/SerializeNVVMTarget.cpp (hunk 2 above)
Comment (joker-eph) [reply mlir/unittests/Target/LLVM/SerializeNVVMTarget.cpp]: Done

**Review by fabianmcg — APPROVED** [review]
Thank you! LGTM, See the formatter error.

File: mlir/lib/IR/BuiltinAttributes.cpp [hunk 4]
```diff
@@ -1544,6 +1544,12 @@ DenseResourceElementsAttr DenseResourceElementsAttr::get(ShapedType type,
   return get(type, manager.insert(blobName, std::move(blob)));
 }
 
+ArrayRef<char> DenseResourceElementsAttr::getData() {
+  if (AsmResourceBlob *blob = this->getRawHandle().getBlob())
+    return blob->template getDataAs<char>();
```
Comment (River707) [inline mlir/lib/IR/BuiltinAttributes.cpp:1549]: Is template necessary here? That should generally only be needed when you're inside of a template function.

**Review by River707 — APPROVED** [review]
My gut wants to say BlobInterface could be a method on ElementsAttr (that only works for contiguous data), but there are some warts there that could make it mildly annoying right now. I'm fine with a BlobInterface right now (given it's very direct), we can come back to ElementsAttr later for cleanup.

**Pull request merged**